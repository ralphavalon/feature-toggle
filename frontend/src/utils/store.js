import React, { useState } from 'react';

export const StoreContext = React.createContext(null);

export default ({ children }) => {
  const [features, setFeatures] = useState([]);
  const [hasError, setHasError] = useState(false);

  const store = {
    features: [features, setFeatures],
    error: [hasError, setHasError]
  };
  return (
    <StoreContext.Provider value={store}>
      {children}
    </StoreContext.Provider>
  );
};

