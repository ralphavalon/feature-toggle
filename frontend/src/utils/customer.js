import React, { useContext } from 'react';
import { StoreContext } from './store';

const getCustomerIds = () => {
  const { features: [features, setFeatures] } = useContext(StoreContext);

  let customerIds = [];
  features.forEach(f => {
    customerIds = customerIds.concat(f.customerIds.filter((item) => customerIds.indexOf(item) < 0));
  });
  return customerIds;
};

export { getCustomerIds };
