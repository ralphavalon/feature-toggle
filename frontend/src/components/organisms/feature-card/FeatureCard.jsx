import React, { useState } from 'react';
import { Form, Row, Col, Button } from 'react-bootstrap';
import Datetime from 'react-datetime';
import PropTypes from 'prop-types';

import FormInputField from '../../molecules/form-input-field';
import FormSelectField from '../../molecules/form-select-field';
import * as Customer from '../../../utils/customer';
import EditAndRemoveHeader from '../../molecules/edit-and-remove-header';
import FormCheckboxField from '../../molecules/form-checkbox-field';

const FeatureCard = ({
  id, isNew, displayName, customerIds,
  description, technicalName, active, expiresOn,
  inverted, onSubmit, onCancel, invalid
}) => {

  const [isEditable, setIsEditable] = useState(isNew);
  const disableEditing = () => !isNew && setIsEditable(false);
  const submitForm = e => {
    onSubmit(e) && disableEditing();
  };
  const cancelForm = () => {
    disableEditing();
    if (onCancel) onCancel();
  };

  return (
    <div>
      <div className='card card-width mb-3'>
        {!isNew && <EditAndRemoveHeader headerText={displayName ? `${id} - ${displayName}` : id}
          onEdit={() => setIsEditable(true)} size={3} className='d-flex justify-content-between' />}
        <div className="card-body">
          <Form noValidate onSubmit={submitForm}>
            {isNew && (
              <FormInputField label="Technical Name" inputType="text" value={technicalName} name="technicalName"
                placeholder="Technical name" className="mb-3" invalidMessage={invalid.technicalName} />
            )}
            {isEditable && (
              <React.Fragment>
                <FormInputField label="Display Name" inputType="text" value={displayName} disabled={!isEditable}
                  name="displayName" placeholder="Display name" className="mb-3" />
                <input type="hidden" name="technicalName" value={technicalName} />
              </React.Fragment>
            )}
            <FormInputField label="Description" inputType="textarea" value={description} name="description" disabled={!isEditable} inputProps={{ rows: 3 }} />
            <FormSelectField label="Customers" name="customerIds" value={customerIds} name="customerIds" className="mb-3"
              disabled={!isEditable} placeholder="Add customers here..." options={Customer.getCustomerIds()}
              invalidMessage={invalid.customerIds} />
            <Form.Group className="mb-3">
              <Form.Label>Expires On</Form.Label>
              <Datetime inputProps={{ name: 'expiresOn', disabled: !isEditable }} initialValue={expiresOn} utc={true} dateFormat="YYYY-MM-DDT" timeFormat="HH:mm:ss" />
            </Form.Group>
            <Row className="mb-3">
              <Col>
                <FormCheckboxField className="text-start" name="active" label="Active" disabled={!isEditable} value={active} />
              </Col>
              <Col>
                <FormCheckboxField className="text-start" name="inverted" label="Inverted" disabled={!isEditable} value={inverted} />
              </Col>
            </Row>
            <Row>
              <Col>
                <Button variant="primary" type="submit" disabled={!isEditable}>Submit</Button>
              </Col>
              {isEditable && (
                <Col>
                  <Button variant="outline" type="button" onClick={cancelForm}>Cancel</Button>
                </Col>)}
            </Row>
          </Form>
        </div>
      </div>
    </div>
  );
};

FeatureCard.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.object.isRequired,
  active: PropTypes.bool,
  customerIds: PropTypes.array,
  description: PropTypes.string,
  displayName: PropTypes.string,
  expiresOn: PropTypes.string,
  id: PropTypes.string,
  inverted: PropTypes.bool,
  isNew: PropTypes.bool,
  onCancel: PropTypes.func,
  technicalName: PropTypes.string,
};

FeatureCard.defaultProps = {
  invalid: {}
};

export default FeatureCard;
