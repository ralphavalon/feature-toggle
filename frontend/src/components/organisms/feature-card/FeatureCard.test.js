import React from 'react';
import { shallow } from 'enzyme';

import * as Customer from '../../../utils/customer';
import FeatureCard from './FeatureCard';

jest.mock('../../../utils/customer');

const mockGetCustomerIds = value => {
  Customer.getCustomerIds.mockImplementation(() => value);
}

beforeEach(() => {
  mockGetCustomerIds([]);
});

afterEach(() => {
  jest.clearAllMocks();
});

const featureCardRequiredProps = {
  onSubmit: jest.fn(),
  onRemove: jest.fn(),
  invalid: {}
};

let featureCardOptionalProps = {
  active: true,
  customerIds: ['123', '321'],
  description: 'description',
  displayName: 'displayName',
  expiresOn: '2021-01-30T11:33:44',
  id: 'id',
  inverted: true,
  isNew: false,
  onCancel: jest.fn(),
  technicalName: 'technicalName',
  invalid: {
    technicalName: 'Error',
    customerIds: 'Other Error',
  }
};

describe('FeatureCard', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<FeatureCard {...featureCardRequiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('renders without disabled because is a new feature', () => {
    const wrapper = shallow(<FeatureCard {...{...featureCardRequiredProps, isNew: true, id: 'id' }}  />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with optional props', () => {
    const wrapper = shallow(<FeatureCard {...{...featureCardRequiredProps, ...featureCardOptionalProps }}  />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should call onCancel after clicking on Cancel', () => {
    const wrapper = shallow(<FeatureCard {...{...featureCardRequiredProps, ...featureCardOptionalProps, isNew: true }}  />);
    expect(featureCardOptionalProps.onCancel).toHaveBeenCalledTimes(0);
    wrapper.find('Button').at(1).simulate('click');
    expect(featureCardOptionalProps.onCancel).toHaveBeenCalledTimes(1);
  });
})

