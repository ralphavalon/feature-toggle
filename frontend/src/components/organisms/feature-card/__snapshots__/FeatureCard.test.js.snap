// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCard renders correctly 1`] = `
<div>
  <div
    className="card card-width mb-3"
  >
    <EditAndRemoveHeader
      className="d-flex justify-content-between"
      onEdit={[Function]}
      size={3}
    />
    <div
      className="card-body"
    >
      <Form
        inline={false}
        noValidate={true}
        onSubmit={[Function]}
      >
        <FormInputField
          disabled={true}
          inputProps={
            Object {
              "rows": 3,
            }
          }
          inputType="textarea"
          label="Description"
          name="description"
        />
        <FormSelectField
          className="mb-3"
          disabled={true}
          label="Customers"
          name="customerIds"
          options={Array []}
          placeholder="Add customers here..."
        />
        <FormGroup
          className="mb-3"
        >
          <FormLabel
            column={false}
            srOnly={false}
          >
            Expires On
          </FormLabel>
          <n
            className=""
            closeOnClickOutside={true}
            closeOnSelect={false}
            closeOnTab={true}
            dateFormat="YYYY-MM-DDT"
            input={true}
            inputProps={
              Object {
                "disabled": true,
                "name": "expiresOn",
              }
            }
            isValidDate={[Function]}
            onBeforeNavigate={[Function]}
            onCalendarClose={[Function]}
            onCalendarOpen={[Function]}
            onChange={[Function]}
            onClose={[Function]}
            onNavigate={[Function]}
            onNavigateBack={[Function]}
            onNavigateForward={[Function]}
            onOpen={[Function]}
            strictParsing={true}
            timeConstraints={Object {}}
            timeFormat="HH:mm:ss"
            utc={true}
          />
        </FormGroup>
        <Row
          className="mb-3"
          noGutters={false}
        >
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={true}
              label="Active"
              name="active"
            />
          </Col>
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={true}
              label="Inverted"
              name="inverted"
            />
          </Col>
        </Row>
        <Row
          noGutters={false}
        >
          <Col>
            <Button
              active={false}
              disabled={true}
              type="submit"
              variant="primary"
            >
              Submit
            </Button>
          </Col>
        </Row>
      </Form>
    </div>
  </div>
</div>
`;

exports[`FeatureCard renders with optional props 1`] = `
<div>
  <div
    className="card card-width mb-3"
  >
    <EditAndRemoveHeader
      className="d-flex justify-content-between"
      headerText="id - displayName"
      onEdit={[Function]}
      size={3}
    />
    <div
      className="card-body"
    >
      <Form
        inline={false}
        noValidate={true}
        onSubmit={[Function]}
      >
        <FormInputField
          disabled={true}
          inputProps={
            Object {
              "rows": 3,
            }
          }
          inputType="textarea"
          label="Description"
          name="description"
          value="description"
        />
        <FormSelectField
          className="mb-3"
          disabled={true}
          label="Customers"
          name="customerIds"
          options={Array []}
          placeholder="Add customers here..."
          value={
            Array [
              "123",
              "321",
            ]
          }
        />
        <FormGroup
          className="mb-3"
        >
          <FormLabel
            column={false}
            srOnly={false}
          >
            Expires On
          </FormLabel>
          <n
            className=""
            closeOnClickOutside={true}
            closeOnSelect={false}
            closeOnTab={true}
            dateFormat="YYYY-MM-DDT"
            initialValue="2021-01-30T11:33:44"
            input={true}
            inputProps={
              Object {
                "disabled": true,
                "name": "expiresOn",
              }
            }
            isValidDate={[Function]}
            onBeforeNavigate={[Function]}
            onCalendarClose={[Function]}
            onCalendarOpen={[Function]}
            onChange={[Function]}
            onClose={[Function]}
            onNavigate={[Function]}
            onNavigateBack={[Function]}
            onNavigateForward={[Function]}
            onOpen={[Function]}
            strictParsing={true}
            timeConstraints={Object {}}
            timeFormat="HH:mm:ss"
            utc={true}
          />
        </FormGroup>
        <Row
          className="mb-3"
          noGutters={false}
        >
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={true}
              label="Active"
              name="active"
              value={true}
            />
          </Col>
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={true}
              label="Inverted"
              name="inverted"
              value={true}
            />
          </Col>
        </Row>
        <Row
          noGutters={false}
        >
          <Col>
            <Button
              active={false}
              disabled={true}
              type="submit"
              variant="primary"
            >
              Submit
            </Button>
          </Col>
        </Row>
      </Form>
    </div>
  </div>
</div>
`;

exports[`FeatureCard renders without disabled because is a new feature 1`] = `
<div>
  <div
    className="card card-width mb-3"
  >
    <div
      className="card-body"
    >
      <Form
        inline={false}
        noValidate={true}
        onSubmit={[Function]}
      >
        <FormInputField
          className="mb-3"
          inputType="text"
          label="Technical Name"
          name="technicalName"
          placeholder="Technical name"
        />
        <FormInputField
          className="mb-3"
          disabled={false}
          inputType="text"
          label="Display Name"
          name="displayName"
          placeholder="Display name"
        />
        <input
          name="technicalName"
          type="hidden"
        />
        <FormInputField
          disabled={false}
          inputProps={
            Object {
              "rows": 3,
            }
          }
          inputType="textarea"
          label="Description"
          name="description"
        />
        <FormSelectField
          className="mb-3"
          disabled={false}
          label="Customers"
          name="customerIds"
          options={Array []}
          placeholder="Add customers here..."
        />
        <FormGroup
          className="mb-3"
        >
          <FormLabel
            column={false}
            srOnly={false}
          >
            Expires On
          </FormLabel>
          <n
            className=""
            closeOnClickOutside={true}
            closeOnSelect={false}
            closeOnTab={true}
            dateFormat="YYYY-MM-DDT"
            input={true}
            inputProps={
              Object {
                "disabled": false,
                "name": "expiresOn",
              }
            }
            isValidDate={[Function]}
            onBeforeNavigate={[Function]}
            onCalendarClose={[Function]}
            onCalendarOpen={[Function]}
            onChange={[Function]}
            onClose={[Function]}
            onNavigate={[Function]}
            onNavigateBack={[Function]}
            onNavigateForward={[Function]}
            onOpen={[Function]}
            strictParsing={true}
            timeConstraints={Object {}}
            timeFormat="HH:mm:ss"
            utc={true}
          />
        </FormGroup>
        <Row
          className="mb-3"
          noGutters={false}
        >
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={false}
              label="Active"
              name="active"
            />
          </Col>
          <Col>
            <FormCheckboxField
              className="text-start"
              disabled={false}
              label="Inverted"
              name="inverted"
            />
          </Col>
        </Row>
        <Row
          noGutters={false}
        >
          <Col>
            <Button
              active={false}
              disabled={false}
              type="submit"
              variant="primary"
            >
              Submit
            </Button>
          </Col>
          <Col>
            <Button
              active={false}
              disabled={false}
              onClick={[Function]}
              type="button"
              variant="outline"
            >
              Cancel
            </Button>
          </Col>
        </Row>
      </Form>
    </div>
  </div>
</div>
`;
